c++/src/kj/common-test.c++

```diff
@@ -793,29 +793,152 @@ TEST(Common, ArrayAsBytes) {
   }
 }
 
-KJ_TEST("ArrayPtr operator ==") {
-  KJ_EXPECT(ArrayPtr<const int>({123, 456}) == ArrayPtr<const int>({123, 456}));
-  KJ_EXPECT(!(ArrayPtr<const int>({123, 456}) != ArrayPtr<const int>({123, 456})));
-  KJ_EXPECT(ArrayPtr<const int>({123, 456}) != ArrayPtr<const int>({123, 321}));
-  KJ_EXPECT(ArrayPtr<const int>({123, 456}) != ArrayPtr<const int>({123}));
-
-  KJ_EXPECT(ArrayPtr<const int>({123, 456}) == ArrayPtr<const short>({123, 456}));
-  KJ_EXPECT(!(ArrayPtr<const int>({123, 456}) != ArrayPtr<const short>({123, 456})));
-  KJ_EXPECT(ArrayPtr<const int>({123, 456}) != ArrayPtr<const short>({123, 321}));
-  KJ_EXPECT(ArrayPtr<const int>({123, 456}) != ArrayPtr<const short>({123}));
-
-  KJ_EXPECT((ArrayPtr<const StringPtr>({"foo", "bar"}) ==
-             ArrayPtr<const char* const>({"foo", "bar"})));
-  KJ_EXPECT(!(ArrayPtr<const StringPtr>({"foo", "bar"}) !=
-             ArrayPtr<const char* const>({"foo", "bar"})));
-  KJ_EXPECT((ArrayPtr<const StringPtr>({"foo", "bar"}) !=
-             ArrayPtr<const char* const>({"foo", "baz"})));
-  KJ_EXPECT((ArrayPtr<const StringPtr>({"foo", "bar"}) !=
-             ArrayPtr<const char* const>({"foo"})));
-
-  // operator== should not use memcmp for double elements.
-  double d[1] = { nan() };
-  KJ_EXPECT(ArrayPtr<double>(d, 1) != ArrayPtr<double>(d, 1));
+enum strong_ordering {
+  strongEqual,
+  strongLess,
+  strongGreater,
+};
+
+enum partial_ordering {
+  unordered,
+  partialEqual,
+  partialLess,
+  partialGreater,
+};
+
+template<typename A, typename B>
+void verifyStrongComparisons(A a, B b, bool expectedEq, bool expectedLT) {
+  KJ_EXPECT((a ==  b) == expectedEq);
+  KJ_EXPECT((b ==  a) == expectedEq);
+  KJ_EXPECT((a !=  b) == !expectedEq);
+  KJ_EXPECT((b !=  a) == !expectedEq);
+  KJ_EXPECT((a <=  b) == (expectedEq || expectedLT));
+  KJ_EXPECT((b <=  a) == !expectedLT);
+  KJ_EXPECT((a >=  b) == !expectedLT);
+  KJ_EXPECT((b >=  a) == (expectedEq || expectedLT));
+  KJ_EXPECT((a <   b) == expectedLT);
+  KJ_EXPECT((b <   a) == !(expectedEq || expectedLT));
+  KJ_EXPECT((a >   b) == !(expectedEq || expectedLT));
+  KJ_EXPECT((b >   a) == expectedLT);
+}
+
+template<typename A, typename B>
+void verifyEqualityComparisons(A a, B b, bool expectedEq) {
+  KJ_EXPECT((a ==  b) == expectedEq);
+  KJ_EXPECT((b ==  a) == expectedEq);
+  KJ_EXPECT((a !=  b) == !expectedEq);
+  KJ_EXPECT((b !=  a) == !expectedEq);
+}
+
+template<typename A, typename B>
+void strongComparisonsTests(A a, B b, strong_ordering ans3way){
+  verifyStrongComparisons<A, B>(a, b,
+    ans3way == strongEqual,
+    ans3way == strongLess);
+}
```

Comment:
should include the real implementation, we have too many levels of abstraction here.